File: Graph/Find if a route exists between 2 nodes in a directed graph.cpp
/*
Given a directed graph, design an algorithm to find out whether there is a route between two nodes.
Note: There could be self loops.
Note: This is a directed graph.
*/
#include <bits/stdc++.h>
using namespace std;
//BFS Traversal
bool hasPath(list<int> adj_list[], bool *visited, int source, int destination)
{
    queue<int> pendingVertex;
    pendingVertex.push(source);
    // visited[source] = true;  to handle cyclic path
    while (!pendingVertex.empty())
    {
        int currentVertex = pendingVertex.front();
        pendingVertex.pop();
        //cout<<currentVertex<<" ";
        for (auto i = adj_list[currentVertex].begin(); i != adj_list[currentVertex].end(); i++)
        {
            if (*i == destination)
                return true;
            if (!visited[*i])
            {
                visited[*i] = true;
                pendingVertex.push(*i);
            }
        }
    }
    return false;
}

bool RouteBetweenNodes(int source, int destination, int numberofnodes, vector<vector<int>> edgelist)
{
    //convert edgeList to adjacency list
    list<int> adj_list[numberofnodes + 1];
    for (int i = 0; i < edgelist.size(); i++)
    {
        int u = edgelist[i][0];
        int v = edgelist[i][1];
        adj_list[u].push_back(v);
    }

    //To keep track of visited nodes
    bool *visited = new bool[numberofnodes + 1];
    for (int i = 0; i <= numberofnodes; i++)
    {
        visited[i] = false;
    }

    bool ans = hasPath(adj_list, visited, source, destination);
    if (ans)
        return true;
    delete[] visited; //delete explicitly
    return false;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int numberofnodes, edges;
        cin >> numberofnodes >> edges;
        int vertex1, vertex2;
        vector<vector<int>> edgelist;
        for (int i = 0; i < edges; i++)
        {
            cin >> vertex1 >> vertex2;
            vector<int> temp = {vertex1, vertex2};
            edgelist.push_back(temp);
        }
        int start, end;
        cin >> start >> end;
        bool result = RouteBetweenNodes(start, end, numberofnodes, edgelist);
        cout << ((result) ? "yes" : "no") << "\n";
    }
    return 0;
}

File: Graph/Find if a route exists between 2 nodes in a directed graph.java
/*
Given a directed graph, design an algorithm to find out whether there is a route between two nodes.
Note: There could be self loops.
Note: This is a directed graph.
*/
import java.util.*;

public class RouteBetweenNodes {
    boolean result;
    int visited[];
    boolean routeBetweenNodes(int source , int destination ,int n,  ArrayList<ArrayList<Integer>> edges) {
        ArrayList<ArrayList<Integer>> graph = new ArrayList<ArrayList<Integer>>();
        for(int i = 0; i <= n;i++){
            graph.add(new ArrayList<Integer>());
        }
        int num_edges = edges.size();
        for(int i = 0;i < num_edges;i++){
            //convert edgelist to adjacency list
            int x = edges.get(i).get(0);
            int y = edges.get(i).get(1);
            graph.get(x).add(y);
        }
        init(n);
        result = false;
        isPath(source,destination,graph);
        return result;
    }
    
    void init(int n){
        this.visited = new int[n + 1];
        for(int i = 0;i <= n;i++){
            this.visited[i]  = 0;
        }
    }

    //DFS Traversal
    void isPath(int source, int destination, ArrayList<ArrayList<Integer>> graph){
        this.visited[source] = 1;
        for(int i = 0; i < graph.get(source).size();i++){
            int node = graph.get(source).get(i);
            if(node == destination){
                this.result = true;
                return;
            }
            if(visited[node] == 0){
                isPath(node,destination,graph); // DFS
            } //can check for success and return if we result == true
        }
        return;
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        for (int p = 0 ; p < t ; p++) {
            int n = scanner.nextInt();
            int m = scanner.nextInt();
            ArrayList<ArrayList<Integer>> edges = new ArrayList<ArrayList<Integer>>();
        
            for(int i=0;i<m;i++){
                edges.add(new ArrayList<Integer>());
                edges.get(i).add(scanner.nextInt());
                edges.get(i).add(scanner.nextInt());
            }
            int src = scanner.nextInt();
            int dest = scanner.nextInt();
            RouteBetweenNodes answer = new RouteBetweenNodes();
            boolean ans = answer.routeBetweenNodes(src, dest,n,edges);
            if (ans) {
                System.out.println("yes");
            } else {
                System.out.println("no");
            }
        }
        scanner.close();
    }
}
